pkg/cache/listers.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+// GenericClusterLister is a lister skin on a generic Indexer
+type GenericClusterLister interface {
+	// List will return all objects across clusters
+	List(selector labels.Selector) (ret []runtime.Object, err error)
+	// ByCluster will give you a GenericLister for one cluster
+	ByCluster(cluster logicalcluster.Name) cache.GenericLister
+}
+
+// NewGenericClusterLister creates a new instance for the genericClusterLister.
+func NewGenericClusterLister(indexer cache.Indexer, resource schema.GroupResource) *genericClusterLister {
+	return &genericClusterLister{indexer: indexer, resource: resource}
+}
+
+type genericClusterLister struct {
+	indexer  cache.Indexer
+	resource schema.GroupResource
+}
+
+func (s *genericClusterLister) List(selector labels.Selector) (ret []runtime.Object, err error) {
+	if selector == nil {
+		selector = labels.NewSelector()
+	}
+	err = cache.ListAll(s.indexer, selector, func(m interface{}) {
+		ret = append(ret, m.(runtime.Object))
+	})
+	return ret, err
+}
+
+func (s *genericClusterLister) ByCluster(cluster logicalcluster.Name) cache.GenericLister {
+	return &genericLister{indexer: s.indexer, resource: s.resource, cluster: cluster}
+}
+
+type genericLister struct {
+	indexer  cache.Indexer
+	cluster  logicalcluster.Name
+	resource schema.GroupResource
+}
+
+func (s *genericLister) List(selector labels.Selector) (ret []runtime.Object, err error) {
+	selectAll := selector == nil || selector.Empty()
```

Comment:
Unit test for everything in this package is needed. Including nil, Everything(), etc